src/psyclone/psyad/transformations/preprocess.py

```diff
@@ -75,3 +82,97 @@ def preprocess_trans(kernel_psyir):
         elif oper.operator == BinaryOperation.Operator.MATMUL:
             # Apply MATMUL transformation
             matmul_trans.apply(oper)
+
+    # Deal with any associativity issues here as AssignmentTrans
+    # is not able to.
+    for assignment in kernel_psyir.walk(Assignment):
+        if assignment.walk(Range):
+            # SymbolicMaths currently does not work if the expression
+            # contains Range nodes, see issue #1655.
+            associativity(assignment, active_variable_names)
+        else:
+            sym_maths = SymbolicMaths.get()
+            sym_maths.expand(assignment.rhs)
+
+
+def associativity(assignment, active_variable_names):
+    '''Repeatedly look for the patterns x * (a +- b) or (a +- b) */ x on
+    the rhs of this assignment where x is an inactive expression and a
+    and b are active expressions, replacing these patterns with x*a +-
+    x*b and a*/x +- b*/x respectively.
+
+    This function can be removed when support for Range nodes is added
+    to thee SymbolicMaths expand function, see issue #1655.
+
+    :param assignment: the Assignment Node that we are looking at.
+    :type assignment: :py:class:`psyclone.psyir.nodes.Assignment`
+    :param active_variable_names: list of active variable names.
+    :type active_variable_names: list of str
+
+    '''
+    # pylint: disable=too-many-boolean-expressions
+    if node_is_active_names(assignment.rhs, active_variable_names):
+        # The rhs of the assignment is active
+        found = True
+        while found:
+            # Repeat until the patterns are no longer found
+            found = False
+            # Check all multiplies on the rhs
+            for oper in assignment.rhs.walk(BinaryOperation):
+                if oper.operator == BinaryOperation.Operator.MUL and \
+                   node_is_passive_names(
+                       oper.children[0], active_variable_names) and \
+                   isinstance(oper.children[1], BinaryOperation) and \
+                   oper.children[1].operator in [
+                       BinaryOperation.Operator.ADD,
+                       BinaryOperation.Operator.SUB] and \
+                   node_is_active_names(
+                       oper.children[1].children[0],
+                       active_variable_names) and \
+                   node_is_active_names(
+                       oper.children[1].children[1],
+                       active_variable_names):
+                    # Matched one of the patterns we are looking for
+                    # x * (a +- b)
+                    found = True
+                    inactive = oper.children[0]
+                    active0 = oper.children[1].children[0]
+                    active1 = oper.children[1].children[1]
+                    binary_op = oper.children[1]
+                    # Restructure to x*a +- x*b
+                    mult0 = BinaryOperation.create(
+                        oper.operator, inactive.detach(), active0.detach())
```

Comment:
Because of the conditional at 122 oper.operator is always going to be `BinaryOperation.Operator.MUL` so I think it would be clearer if you used that here and for `mult1` too.

Reply:
OK